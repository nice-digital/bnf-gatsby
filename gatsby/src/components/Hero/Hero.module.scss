@use '../../styles/colours';
@use '@nice-digital/nds-core/scss/layout';
@use '@nice-digital/nds-core/scss/media-queries';
@use '@nice-digital/nds-core/scss/spacing';
@use '@nice-digital/nds-core/scss/typography';
@use '@nice-digital/nds-core/scss/utils';

.hero {
  background: colours.$hero-background;
  left: 50%;
  margin-bottom: utils.rem(spacing.$large);
  margin-left: -50vw;
  margin-right: -50vw;
  overflow: hidden;
  padding-bottom: utils.rem(spacing.$large);
  position: relative;
  right: 50%;
  width: 100vw;

  .heroContainer {
    @include layout.container;
  }

  .title {
    @include typography.h1;
  }

  .intro {
    @include typography.lead;
    max-width: unset;
  }

  .lastUpdated {
    h2 {
      @include typography.h5;
      margin: 0;
    }

    time {
      @include typography.h3;
      display: block;
      font-weight: 700;
      margin: 0 0 utils.rem(spacing.$medium);
    }
  }
}

.breadcrumbs {
  flex: 0 1 100%;
}

@include media-queries.mq($from: md) {
  .heroContainer {
    align-items: center;
    column-gap: utils.rem(spacing.$medium);
    display: flex;
    flex-flow: row;
    flex-wrap: wrap;
  }

  .text {
    flex-grow: 1;
  }

  .lastUpdated {
    background-color: colours.$hero-last-updated-background;
    padding: utils.rem(spacing.$medium);
    padding-left: utils.rem(spacing.$medium);
    padding-right: utils.rem(spacing.$large);
  }

  .lastUpdated * {
    white-space: nowrap;
  }
}
