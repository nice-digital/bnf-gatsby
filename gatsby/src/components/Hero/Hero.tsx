import React from "react";

import styles from "./Hero.module.scss";

type HeroProps = {
	id: string;
	title: string;
	intro: string;
};

export const Hero: React.FC<HeroProps> = (props: HeroProps) => {
	const { id, title, intro } = props;
	return (
		<div className={styles.hero} id={id}>
			<div className={styles.hero_container}>
				<div className={styles.hero__body}>
					<div className={styles.hero__copy}>
						<h1 className={styles.hero__title}>{title}</h1>
						{intro && <p className={styles.hero__intro}>{intro}</p>}
					</div>
				</div>
			</div>
		</div>
	);
};
