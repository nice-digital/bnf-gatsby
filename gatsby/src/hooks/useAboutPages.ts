import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface AboutPages {
	allBnfAboutSection: {
		aboutSectionPages: SlugAndTitle[];
	};
	allBnfCautionaryAndAdvisoryGuidance: {
		labelsGuidancePages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/about/${slug}/`,
});

/**
 * Gets the lists of pages for the about section by combining:
 * - about sections
 * - cautionary/advisory labels
 * - cautionary/advisory guidance page(s)
 * from the feed.
 *
 * @returns The consolidated list of about pages
 */
export const useAboutPages = (): MenuPageLink[] => {
	const { allBnfAboutSection, allBnfCautionaryAndAdvisoryGuidance } =
		useStaticQuery<AboutPages>(
			graphql`
				query AboutPages {
					allBnfAboutSection(sort: { order: ASC }) {
						aboutSectionPages: nodes {
							slug
							title
						}
					}
					allBnfCautionaryAndAdvisoryGuidance {
						labelsGuidancePages: nodes {
							title
							slug
						}
					}
				}
			`
		);

	return [
		...allBnfAboutSection.aboutSectionPages.map(slugToHref),
		{ title: "Labels", href: "/about/labels/" },
		...allBnfCautionaryAndAdvisoryGuidance.labelsGuidancePages.map(slugToHref),
	];
};
