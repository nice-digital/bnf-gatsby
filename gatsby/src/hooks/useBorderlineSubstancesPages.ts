import { useStaticQuery, graphql } from "gatsby";

import { SID, PHPID } from "@nice-digital/gatsby-source-bnf";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface BorderlineSubstancesQueryResult {
	allBnfBorderlineSubstancesTaxonomyRoot: {
		pages: {
			slug: string;
			title: string;
		}[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/borderline-substances/${slug}/`,
});

/**
 * Gets the lists of pages for the borderline substances section
 *
 * @returns The consolidated list of borderline substances pages
 */
export const useBorderlineSubstancesPages = (): MenuPageLink[] => {
	const { allBnfBorderlineSubstancesTaxonomyRoot } =
		useStaticQuery<BorderlineSubstancesQueryResult>(
			graphql`
				{
					allBnfBorderlineSubstancesTaxonomyRoot {
						pages: nodes {
							slug
							title
						}
					}
				}
			`
		);

	return allBnfBorderlineSubstancesTaxonomyRoot.pages.map(slugToHref);
};
