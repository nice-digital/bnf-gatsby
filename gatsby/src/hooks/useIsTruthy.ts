import { useCallback } from "react";

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useIsTruthy = () =>
	useCallback(
		<TMaybe>(maybeT: TMaybe | null): maybeT is TMaybe => !!maybeT,
		[]
	);
