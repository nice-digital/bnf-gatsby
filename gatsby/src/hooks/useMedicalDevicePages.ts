import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface MedicalDevicesQueryResult {
	allBnfMedicalDevice: {
		pages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/medical-devices/${slug}/`,
});

/**
 * Gets the lists of pages for the medical devices section
 *
 * @returns The consolidated list of medical devices pages
 */
export const useMedicalDevicePages = (): MenuPageLink[] => {
	const { allBnfMedicalDevice } = useStaticQuery<MedicalDevicesQueryResult>(
		graphql`
			{
				allBnfMedicalDevice(sort: { fields: title, order: ASC }) {
					pages: nodes {
						slug
						title
					}
				}
			}
		`
	);

	return allBnfMedicalDevice.pages.map(slugToHref);
};
