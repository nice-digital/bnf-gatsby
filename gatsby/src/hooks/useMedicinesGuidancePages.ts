import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface MedicinesGuidanceQueryResult {
	allBnfGuidance: {
		pages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/medicines-guidance/${slug}/`,
});

/**
 * Gets the lists of pages for the medicines guidance section
 *
 * @returns The consolidated list of medicines guidance pages
 */
export const useMedicinesGuidancePages = (): MenuPageLink[] => {
	const { allBnfGuidance } = useStaticQuery<MedicinesGuidanceQueryResult>(
		graphql`
			{
				allBnfGuidance(sort: { fields: order, order: ASC }) {
					pages: nodes {
						slug
						title
					}
				}
			}
		`
	);

	return allBnfGuidance.pages.map(slugToHref);
};
