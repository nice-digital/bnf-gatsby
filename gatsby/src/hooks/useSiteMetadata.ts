import { useStaticQuery, graphql } from "gatsby";

interface SiteMetaData {
	/** Whether the current site is BNF or BNFC */
	isBNF: boolean;
	/** Absolute URL to the site homepage, without trailing slash */
	siteUrl: string;
	/** Short site title e.g. "BNF" or "BNFC" */
	siteTitleShort: string;
	/** Full site title e.g. "British National Formulary" or "British National Formulary for Children" */
	siteTitleLong: string;
}

interface Site {
	site: {
		siteMetadata: SiteMetaData;
	};
}

/**
 * Gets the site metadata, as defined in gatsby-config
 *
 * @returns The site metadata
 */
export const useSiteMetadata = (): SiteMetaData => {
	const { site } = useStaticQuery<Site>(
		graphql`
			query SiteMetaData {
				site {
					siteMetadata {
						isBNF
						siteUrl
						siteTitleShort
						siteTitleLong
					}
				}
			}
		`
	);
	return site.siteMetadata;
};
