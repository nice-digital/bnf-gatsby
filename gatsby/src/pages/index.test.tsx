import { render, screen } from "@testing-library/react";
import { useStaticQuery } from "gatsby";

import { type LastUpdatedDataQueryResult } from "@/components/Hero/Hero";
import { useSiteMetadata } from "@/hooks/useSiteMetadata";

import HomePage from "./";

export const mockLastUpdatedDataQueryData: LastUpdatedDataQueryResult = {
	bnfMetadata: {
		lastUpdatedDate: "2022-04-06T00:38:57.911",
		lastUpdatedDateFormatted: "6 April 2022",
		runTag: "2022_04_06_0037_bnf",
	},
};

(useStaticQuery as jest.Mock).mockReturnValue(mockLastUpdatedDataQueryData);

describe("HomePage", () => {
	it("should match snapshot for BNF", () => {
		render(<HomePage />);
		expect(screen.getByRole("main")).toMatchSnapshot();
	});

	it("should match snapshot for BNFC", () => {
		(useSiteMetadata as jest.Mock).mockReturnValueOnce({
			isBNF: false,
			siteTitleShort: "BNFC",
			siteTitleLong: "British National Formulary for Children",
		});

		render(<HomePage />);
		expect(screen.getByRole("main")).toMatchSnapshot();
	});

	describe("Drug list", () => {
		it("should render an A-Z link for all letters", () => {
			render(<HomePage />);
			expect(screen.queryByText("A")?.tagName).toBe("A");
			expect(screen.queryByText("B")?.tagName).toBe("A");
			expect(screen.queryByText("D")?.tagName).toBe("A");
			expect(screen.queryByText("K")?.tagName).toBe("A");
			expect(screen.queryByText("Z")?.tagName).toBe("A");
		});

		it("should render letters that link to /drugs/ with appropriate anchor", () => {
			render(<HomePage />);
			const linkElementA = screen.getByText("A");
			expect(linkElementA).toHaveAttribute("href", "/drugs/#a");

			const linkElementH = screen.getByText("H");
			expect(linkElementH).toHaveAttribute("href", "/drugs/#h");

			const linkElementZ = screen.getByText("Z");
			expect(linkElementZ).toHaveAttribute("href", "/drugs/#z");
		});
	});

	describe("Headings", () => {
		it.each([
			["Drugs", 2],
			["Browse drugs by A-Z", 3],
			["Treatment summaries", 2],
			["Interactions", 2],
			["Medicines guidance", 2],
			["Wound management", 2],
			["Medical devices", 2],
			["Borderline substances", 2],
			["Nurse prescribers formulary", 2],
			["Dental practitioners formulary", 2],
			["Approximate conversions and units", 2],
			["Cautionary and advisory labels", 2],
			["Abbreviations and symbols", 2],
		])("should render a '%s' level %d heading", (name, level) => {
			render(<HomePage />);

			expect(screen.getByRole("heading", { name, level })).toBeInTheDocument();
		});
	});

	describe("Wound management rendering per site", () => {
		it("should render wound management section for BNF ", () => {
			render(<HomePage />);

			expect(
				screen.getByRole("link", {
					name: "Wound management",
				})
			).toHaveAttribute("href", "/wound-management/");
		});

		it("should not render wound management section for BNFC query by role", () => {
			(useSiteMetadata as jest.Mock).mockReturnValueOnce({
				isBNF: false,
				siteTitleShort: "BNFC",
				siteTitleLong: "British National Formulary for Children",
			});

			render(<HomePage />);

			expect(
				screen.queryByRole("link", { name: "Wound management" })
			).toBeNull();
		});
	});
});
