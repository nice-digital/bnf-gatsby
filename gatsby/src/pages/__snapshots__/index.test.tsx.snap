// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should match snapshot for BNF 1`] = `
<main>
  <div
    class="container"
  >
    <div
      class="hero"
      id="content-start"
    >
      <div
        class="hero__container"
      >
        <nav
          aria-label="Breadcrumbs"
          role="navigation"
        >
          <ol
            class="breadcrumbs"
          >
            <li
              class="breadcrumbs__crumb"
            >
              <a
                href="https://www.nice.org.uk/"
              >
                NICE
              </a>
            </li>
            <li
              class="breadcrumbs__crumb"
            >
              <span>
                BNF
              </span>
            </li>
          </ol>
        </nav>
        <script
          type="application/ld+json"
        >
          {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "https://www.nice.org.uk/",
        "name": "NICE"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "name": "BNF"
      }
    }
  ]
}
        </script>
        <div
          class="hero__body"
        >
          <div
            class="hero__copy"
          >
            <h1
              class="hero__title"
            >
              British National Formulary
            </h1>
            <p
              class="hero__intro"
            >
              The British National Formulary (BNF) is the first choice for concise medicines information. Trusted by healthcare professionals across the world to support confident decision-making at the point of care.
            </p>
            <div
              class="hero__actions"
            >
              <a
                class="btn btn--cta"
                href="/drugs/"
              >
                Browse drugs
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="grid grid--loose"
    >
      <div
        data-g="12 sm:6 md:4"
      >
        <div
          class="panel panel--impact-alt"
        >
          <h2
            class="h3"
          >
            Drugs A to Z
          </h2>
          <p>
            Browse the list of drug monographs, arranged alphabetically.
          </p>
          <a
            class="btn btn--cta"
            href="/drugs/"
          >
            Browse drugs
          </a>
        </div>
      </div>
      <div
        data-g="12 sm:6 md:4"
      >
        <div
          class="panel panel--primary"
        >
          <h2
            class="h3"
          >
            Interactions
          </h2>
          <p>
            Browse the list of drug interactions, arranged alphabetically.
          </p>
          <a
            class="btn btn--inverse"
            href="/interactions/"
          >
            Browse interactions
          </a>
        </div>
      </div>
      <div
        data-g="12 sm:6 md:4"
      >
        <div
          class="panel panel--impact-alt"
        >
          <h2
            class="h3"
          >
            Treatment summaries
          </h2>
          <p>
            Browse the list of treatment summaries, arranged alphabetically.
          </p>
          <a
            class="btn btn--secondary"
            href="/treatment-summaries/"
          >
            Browse treatment summaries
          </a>
        </div>
      </div>
    </div>
    <p>
      <a
        href="/about/"
      >
        About
      </a>
    </p>
    <p>
      <a
        href="/dental-practitioners-formulary/"
      >
        Dental Practitioners' Formulary
      </a>
    </p>
    <p>
      <a
        href="/medical-devices/"
      >
        Medical devices
      </a>
    </p>
    <p>
      <a
        href="/wound-management/"
      >
        Wound management products and elasticated garments
      </a>
    </p>
    <p>
      <a
        href="/nurse-prescribers-formulary/"
      >
        Nurse Prescribers' Formulary
      </a>
    </p>
  </div>
</main>
`;

exports[`HomePage should match snapshot for BNFC 1`] = `
<main>
  <div
    class="container"
  >
    <div
      class="hero"
      id="content-start"
    >
      <div
        class="hero__container"
      >
        <nav
          aria-label="Breadcrumbs"
          role="navigation"
        >
          <ol
            class="breadcrumbs"
          >
            <li
              class="breadcrumbs__crumb"
            >
              <a
                href="https://www.nice.org.uk/"
              >
                NICE
              </a>
            </li>
            <li
              class="breadcrumbs__crumb"
            >
              <span>
                BNF
              </span>
            </li>
          </ol>
        </nav>
        <script
          type="application/ld+json"
        >
          {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "https://www.nice.org.uk/",
        "name": "NICE"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "name": "BNF"
      }
    }
  ]
}
        </script>
        <div
          class="hero__body"
        >
          <div
            class="hero__copy"
          >
            <h1
              class="hero__title"
            >
              British National Formulary for Children
            </h1>
            <p
              class="hero__intro"
            >
              Covering neonates to adolescents, the BNF for Children includes key clinical and pharmaceutical information specific to those age groups.
            </p>
            <div
              class="hero__actions"
            >
              <a
                class="btn btn--cta"
                href="/drugs/"
              >
                Browse drugs
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="grid grid--loose"
    >
      <div
        data-g="12 sm:6 md:4"
      >
        <div
          class="panel panel--impact-alt"
        >
          <h2
            class="h3"
          >
            Drugs A to Z
          </h2>
          <p>
            Browse the list of drug monographs, arranged alphabetically.
          </p>
          <a
            class="btn btn--cta"
            href="/drugs/"
          >
            Browse drugs
          </a>
        </div>
      </div>
      <div
        data-g="12 sm:6 md:4"
      >
        <div
          class="panel panel--primary"
        >
          <h2
            class="h3"
          >
            Interactions
          </h2>
          <p>
            Browse the list of drug interactions, arranged alphabetically.
          </p>
          <a
            class="btn btn--inverse"
            href="/interactions/"
          >
            Browse interactions
          </a>
        </div>
      </div>
      <div
        data-g="12 sm:6 md:4"
      >
        <div
          class="panel panel--impact-alt"
        >
          <h2
            class="h3"
          >
            Treatment summaries
          </h2>
          <p>
            Browse the list of treatment summaries, arranged alphabetically.
          </p>
          <a
            class="btn btn--secondary"
            href="/treatment-summaries/"
          >
            Browse treatment summaries
          </a>
        </div>
      </div>
    </div>
    <p>
      <a
        href="/about/"
      >
        About
      </a>
    </p>
    <p>
      <a
        href="/dental-practitioners-formulary/"
      >
        Dental Practitioners' Formulary
      </a>
    </p>
    <p>
      <a
        href="/medical-devices/"
      >
        Medical devices
      </a>
    </p>
    <p>
      <a
        href="/wound-management/"
      >
        Wound management products and elasticated garments
      </a>
    </p>
    <p>
      <a
        href="/nurse-prescribers-formulary/"
      >
        Nurse Prescribers' Formulary
      </a>
    </p>
  </div>
</main>
`;
